internal/storage/download_test.go

```diff
@@ -145,49 +208,97 @@ func TestDownloadToFileError(t *testing.T) {
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err != nil {
 		t.Fatalf("failed to download file artifact: %v", err)
 	}
 	check(name, art.Size, t)
 
 	// 2. Try with missing checksum.
 	art.HashValue = ""
-	if err := downloadArtifact(name, art, nil, make(chan struct{})); err == nil {
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
 		t.Fatal("validatted with missing checksum")
 	}
 
 	// 3. Try with missing link.
 	art.Link = "http://localhost:43234/test-missing.txt"
-	if err := downloadArtifact(name, art, nil, make(chan struct{})); err == nil {
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
 		t.Fatal("failed to validate with missing link")
 	}
 
 	// 4. Try with wrong checksum type.
 	art.Link = "http://localhost:43234/test-simple.txt"
 	art.HashType = ""
-	if err := downloadArtifact(name, art, nil, make(chan struct{})); err == nil {
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
 		t.Fatal("validate with wrong checksum type")
 	}
 
 	// 5. Try with wrong checksum format.
 	art.HashValue = ";;"
-	if err := downloadArtifact(name, art, nil, make(chan struct{})); err == nil {
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
 		t.Fatal("validate with wrong checksum format")
 	}
 
-	// 5. Try to download file bigger than expected.
+	// 6. Try to download file bigger than expected.
 	art.HashType = "MD5"
 	art.HashValue = "ab2ce340d36bbaafe17965a3a2c6ed5b"
 	art.Size -= 10
-	if err := downloadArtifact(name, art, nil, make(chan struct{})); err == nil {
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
 		t.Fatal("validate with file bigger than expected")
 	}
+
+}
+
+// TestDownloadToFileSecureError tests HTTPS file download function for bad/expired TLS certificates.
+func TestDownloadToFileSecureError(t *testing.T) {
+	// Prepare
+	dir := "_tmp-download"
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		t.Fatalf("failed create temporary directory: %v", err)
+	}
+
+	// Remove temporary directory at the end
+	defer os.RemoveAll(dir)
+
+	art := &Artifact{
+		FileName: "test.txt", Size: 65536, Link: "https://localhost:43234/test.txt",
+		HashType:  "MD5",
+		HashValue: "ab2ce340d36bbaafe17965a3a2c6ed5b",
+	}
+
+	// Start Web servers secure
+	srvSecureInv := host(":43234", art.FileName, int64(art.Size), true, true, expiredCert, expiredKey, t)
+	defer srvSecureInv.close()
+	srvSecureExt := host(":43235", art.FileName, int64(art.Size), true, true, extCert, extKey, t)
+	defer srvSecureExt.close()
+	srvSecureVal := host(":43236", art.FileName, int64(art.Size), true, true, validCert, validKey, t)
+	defer srvSecureVal.close()
+	name := filepath.Join(dir, art.FileName)
+
+	// 1. Try download with expired certificate
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
+		t.Fatalf("validated with expired certificate: %v", err)
```

Comment:
The message is not absolutely correct, the client is not configured with the server certificate.